qualtran/qir_interop/bloq_to_qir.py

```diff
@@ -0,0 +1,349 @@
+
+    Args:
+        bloq: Bloq to get number of qubits from.
+
+    Returns:
+        The number of qubits used in the given Bloq
+    """
+    num_qubits = 0
+    for register in bloq.signature.lefts():
+        shape = register.shape[0] if len(register.shape) != 0 else 1
+        num_qubits += register.bitsize * shape
+    return num_qubits
+
+
+def create_func_for_bloq(bloq: qualtran.Bloq, name: str, qubit_type: Type, void_type,
+                         mod: pyqir.Module) -> Function:
+    """
+    Create a QIR function for the given Bloq.
+
+    Args:
+        bloq: Bloq to create the QIR function for.
+        name: Name of the QIR function.
+        qubit_type: Qubit type for the QIR function.
+        void_type: Void type for the QIR function.
+        mod: Module to add the QIR function to.
+    Returns:
+        The QIR function for the given Bloq.
+    """
+    num_qubits = get_num_qubits_for_bloq(bloq)
+    return Function(
+        FunctionType(void_type, [qubit_type] * num_qubits),
+        Linkage.EXTERNAL,
+        name,
+        mod
+    )
+
+
+def create_ir_map(bloq: qualtran.Bloq) -> dict:
+    """
+    Creates a dictionary which maps the IR of qubits to an index.
+
+    Given that we have a QIR function with a list of parameters, params,
+    and len(params) = m:
+    The IR map of a Qualtran Bloq register named 'x' with n qubits is a map
+    with entries ('x', i): j where i goes from 0 to n - 1 and 0<=j<m.
+    The IR map of the entire Qualtran Bloq is the union of the IR of each register in the Bloq.
+    Note that for all j in {0,1,..., m - 1}, there should exist a key such that IR[key] = j.
+
+    This function creates such a map.
+
+    Args:
+        bloq: Bloq to create the IR map for.
+
+    Returns:
+        Dictionary which maps the IR of qubits to an index.
+    """
+    param_counter = 0
+    ir_map = {}
+    # Loop through all registers in signature
+    for register in bloq.signature.lefts():
+        shape = register.shape[0] if len(
+            register.shape) != 0 else 1  # get the shape as an int (we are asumming its 1d for simplicity)
+        # map the (register_name, index_in_register) to the overall index
+        ir_map.update(
+            {(register.name, i * register.bitsize + j): param_counter + i * register.bitsize + j for
+             i in range(shape) for j in range(register.bitsize)})
+        param_counter += register.bitsize * shape
+    return ir_map
+
+
+def find_ir_for_index(i: int, ir_map: dict) -> Tuple[str, int]:
+    """
+    Given an index i, find the IR for that index.
+
+    Args:
+        i: Index to find the IR for.
+        ir_map: The IR map to use.
+
+    Returns:
+        The IR for the given index.
+    """
+    for key, val in ir_map.items():
+        if i == val:
+            return key
+    raise Exception(f"Error in find_ir_for_index: No IR for index {i}")
+
+
+def cz_power(angle: float, params: List[pyqir._native.Constant], builder, qubit_alloc):
```

Comment:
@tanujkhattar I may have misheard you in the meeting when you said the decomposition of cz power into toffoli + power gate. Please let me know if this is incorrect.